File: etc/api/BaseX.php
<?php
class Session {
	
	// Constructor.
	function __construct($h, $p, $user, $pw) {
		global $socket;
		$socket = socket_create(AF_INET, SOCK_STREAM, 0);
		$result = socket_connect($socket, $h, $p); 
    	if (!$result) {
     		throw new Exception("Can't communicate with the server.");
    	}
    	else {
    		// receive timestamp
   			$ts = $this->readString();
    		$pw = hash("md5", $pw);
    		$all = $pw.$ts;
    		$end = hash("md5", $all);
    
    		// send username and hashed password/timestamp
    		socket_write($socket, $user);
    		socket_write($socket, "\x00");
    		socket_write($socket, $end);
    		socket_write($socket, "\x00");
    
    	// receives success flag
    	if (socket_read($socket, 1) != "\x00") {
    		throw new Exception("Access denied.");
    	}	
	  }		
	}
	
	// Executes a command.
	public function execute($com, $out) {
		global $socket;
		global $info;
		
		socket_write($socket, $com);
		socket_write($socket, "\x00");
		fwrite($out, $this->readString());
		$info = $this->readString();
		return socket_read($socket, 1);
	}
	
	// Returns the info string.
	public function info() {
		global $info;
		return $info;
	}
	
	// Initiates the incoming message.
	private function init() {
		global $message;
		global $pos;
		$message = "";
		$pos = 0;
	}
	
	// Receives a string from the socket.
	private function readString() {
		global $socket;
    	$com = "";
    	while (($data = socket_read($socket, 1)) != "\x00") {
    		$com = $com.$data;
    	}
    	return $com;
	}
	
	// Closes the socket.
	public function closeSocket() {
		global $socket;
		socket_close($socket);
	} 	
}

class Client {
	
	// Constructor.
	function __construct($h, $p) {
		global $host;
		global $port;
		$host = $h;
		$port = $p;
	}
	
	// Creates a session.
	function session() {
		global $host;
		global $port;
		$username = "admin";
		$password = "admin";
		try {
			$session = new Session($host, $port, $username, $password);
		} catch (Exception $e) {
			echo $e->getMessage();
		}
	}
}

$host = "localhost";
$port = 1984;
$client = new Client($host, $port);
$client->session();
?>

File: etc/api/BaseX.py
#
# This Python Module provides two classes for connecting to the
# BaseX Server.
#
# The Client-Class is a standard client which initiates the connection
# to the server, listens to user input and shows the server output.
#
# The Session-Class manages the communication between the server and the client.
# This class has to be used for creating your own client (see Example.py).
#
# The Constructor of the Session-Class needs a hostname, port, username and
# password for the connection. The socket connection will then be established via
# the hostname and the port.
#
# For the execution of commands you need to specify an output stream to pass it
# to the execution method.
#
# (C) Workgroup DBIS, University of Konstanz 2005-10, ISC License

import hashlib, socket, array, getopt, sys, getpass

class Session():

  # Initializes the session.
  def __init__(self, host, port, user, pw):
    global s
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.connect((host, port))

    # allocate 4kb buffer
    self.__buf = array.array('B', '\0' * 0x1000)
    self.__init()

    # receive timestamp
    ts = self.__readString()

    pwmd5 = hashlib.md5(pw).hexdigest()
    m = hashlib.md5()
    m.update(pwmd5)
    m.update(ts)
    complete = m.hexdigest()

    # send user name and hashed password/timestamp
    s.send(user)
    s.send('\0')
    s.send(complete)
    s.send('\0')

    # receives success flag
    if s.recv(1) != '\0':
      raise IOError("Access Denied.")

  # Executes a command.
  def execute(self, com, out):
    s.send(com)
    s.send('\0')

    self.__init()
    out.write(self.__readString())
    self.__info = self.__readString()
    return self.__read() == 0

  # Returns the info string.
  def info(self):
    return self.__info

  # Closes the socket.
  def close(self):
    s.send('exit')
    s.close()

  # Initializes the byte transfer
  def __init(self):
    self.__bpos = 0
    self.__bsize = 0

  # Receives a string from the socket.
  def __readString(self):
    bf = array.array('B')
    while True:
      b = self.__read()
      if b:
        bf.append(b)
      else:
        return bf.tostring()

  # Returns the next byte
  def __read(self):
    # Cache next bytes
    if self.__bpos == self.__bsize:
      self.__bsize = s.recv_into(self.__buf)
      self.__bpos = 0

    # Return current byte
    b = self.__buf[self.__bpos]
    self.__bpos += 1
    return b

# This class offers an interactive BaseX console
class Client(object):
  # Initializes the client.
  def __init__(self, host, port):
    self.__host = host
    self.__port = port
    print 'BaseX Client'
    print 'Try "help" to get some information.'

  # Creates a session.
  def session(self):
    user = raw_input('Username: ');
    pw = getpass.getpass('Password: ');
    try:
      global session
      session = Session(self.__host, self.__port, user, pw)
      self.__console()
      session.close()
      print "See you."
    except IOError as e:
      print e

  # Runs the console.
  def __console(self):
    out = sys.stdout
    session.execute("SET INFO ON", out)
    while True:
      com = str(raw_input('> ')).strip()
      if com == 'exit':
        break
      if com:
        session.execute(com, out)
        print session.info()

# Reads arguments -p and -h.
def opts():
    try:
      opts, args = getopt.getopt(sys.argv[1:], "-p:-h", ["port", "host"])
    except getopt.GetoptError, err:
      print str(err)
      sys.exit()
    global host
    global port
    host = "localhost"
    port = 1984

    for o, a in opts:
      if o == "-p":
        port = int(a)
      if o == "-h":
        host = a

# Main method.
if __name__ == '__main__':
  opts()
  Client(host, port).session()
